File: core-android/src/main/java/com/lookaround/core/android/camera/SingleThreadHandlerExecutor.java
package com.lookaround.core.android.camera;

import android.os.Handler;
import android.os.HandlerThread;

import androidx.annotation.NonNull;

import java.util.concurrent.Executor;
import java.util.concurrent.RejectedExecutionException;

final class SingleThreadHandlerExecutor implements Executor {
    private final String threadName;
    private final HandlerThread handlerThread;
    private final Handler handler;

    public SingleThreadHandlerExecutor(@NonNull String threadName, int priority) {
        this.threadName = threadName;
        handlerThread = new HandlerThread(threadName, priority);
        handlerThread.start();
        handler = new Handler(handlerThread.getLooper());
    }

    @NonNull
    public Handler getHandler() {
        return handler;
    }

    @Override
    public void execute(@NonNull Runnable command) {
        if (!handler.post(command)) {
            throw new RejectedExecutionException(threadName + " is shutting down.");
        }
    }

    public boolean shutdown() {
        return handlerThread.quitSafely();
    }
}

File: core-android/src/main/java/com/lookaround/core/android/camera/SingleThreadHandlerExecutor.kt
package com.lookaround.core.android.camera

import android.os.Handler
import android.os.HandlerThread
import java.util.concurrent.Executor
import java.util.concurrent.RejectedExecutionException

internal class SingleThreadHandlerExecutor(private val threadName: String, priority: Int) :
    Executor {
    private val handlerThread: HandlerThread = HandlerThread(threadName, priority)
    val handler: Handler

    init {
        handlerThread.start()
        handler = Handler(handlerThread.looper)
    }

    override fun execute(command: Runnable) {
        if (!handler.post(command)) {
            throw RejectedExecutionException("$threadName is shutting down.")
        }
    }

    fun shutdown(): Boolean = handlerThread.quitSafely()
}
